src/cpu/pred/bpred_unit.hh

```diff
@@ -213,88 +248,133 @@ class BPredUnit : public SimObject
         {
             return this->seqNum == entry.seqNum;
         }
 
         /** The sequence number for the predictor history entry. */
-        InstSeqNum seqNum;
+        const InstSeqNum seqNum;
 
-        /** The PC associated with the sequence number. */
-        Addr pc;
+        /** The thread id. */
+        const ThreadID tid;
 
-        /** Pointer to the history object passed back from the branch
-         * predictor.  It is used to update or restore state of the
-         * branch predictor.
-         */
-        void *bpHistory = nullptr;
+        /** The PC associated with the sequence number. */
+        const Addr pc;
 
-        void *indirectHistory = nullptr;
+        /** The branch instrction */
+        const StaticInstPtr inst;
 
-        void *rasHistory = nullptr;
+        /** The type of the branch */
+        const BranchType type;
 
-        /** The RAS target (only valid if a return). */
-        std::unique_ptr<PCStateBase> RASTarget;
+        /** Whether or not the instruction was a call. */
+        const bool call;
 
-        /** The RAS index of the instruction (only valid if a call). */
-        unsigned RASIndex = 0;
-
-        /** The thread id. */
-        ThreadID tid;
+        /** Was unconditional control */
+        const bool uncond;
 
         /** Whether or not it was predicted taken. */
         bool predTaken;
 
-        /** Whether or not the RAS was used. */
-        bool usedRAS = false;
+        /** To record the actual outcome of the branch */
+        bool actuallyTaken;
+
+        /** The prediction of the conditional predictor */
+        bool condPred;
+
+        /** Was BTB hit at prediction time */
+        bool btbHit;
 
-        /* Whether or not the RAS was pushed */
-        bool pushedRAS = false;
+        /** Which component provided the target */
+        TargetProvider targetProvider;
 
-        /** Wether this instruction was an indirect branch */
-        bool wasIndirect = false;
+        /** Resteered */
+        bool resteered;
 
-        /** Target of the branch. First it is predicted, and fixed later
-         *  if necessary
+        /** The branch was corrected hence was mispredicted. */
+        bool mispredict;
+
+        /** The predicted target */
+        std::unique_ptr<PCStateBase> target;
+
+        /**
+         * Pointer to the history objects passed back from the branch
+         * predictor subcomponents.
+         * It is used to update or restore state.
+         * Respectively for conditional, indirect and RAS.
          */
-        Addr target = MaxAddr;
+        void *bpHistory = nullptr;
 
-        /** The branch instrction */
-        const StaticInstPtr inst;
+        void *indirectHistory = nullptr;
+
+        void *rasHistory = nullptr;
 
-        /** Whether this branch was mispredicted */
-        bool mispredict = false;
     };
 
-    typedef std::deque<PredictorHistory> History;
+    typedef std::deque<PredictorHistory*> History;
```

Comment:
I just kept it as it was but since I'm anyway refactoring we can consider renaming.

The `PredictorHistory` basically captures the state at prediction time to be able to roll back.
And the `History` is a buffer of outstanding not yet committed predictions

Considering that I would say it should be `PredictionInfo` or `PredictInfo`.
The history "queue" could be `PredictQueue` or `PredictionBuffer`.
However, if I'm correct Intel calls it Branch-order-buffer` and I have seen that terminology at some point as well.

If that is ok then I'll change the naming to: 
`PredictorHistory` -> `PredictInfo`.
The `predHist` -> `bob` for branch-order-buffer.

Also, then I should change all`hist` to `pi` for an instance of `PredictInfo`

Reply:
Ah, sorry, I didn't realize that was an existing name from the previous code.
We can leave it as it is if you like. but I did find it a little confusing having a `History` being a collection of `PredictorHistory`s. I see what you mean, though - the `PredictorHistory` also contains histories.

Considering the alternatives, perhaps it's better to leave things as they are. @andysan do you have an opinion on this?

Reply:
I would suggest not changing it in this PR given that it is the old name. No need to make this unnecessarily complicated, I would rather merge this as soon as possible and focus on the new functionality.